Make postgres index reflection more robust

markstory opened the pull request:
Use postgres' catalog schema more to allow complex index definitions like:

```
CREATE INDEX flags_unaccent_title_idx ON flags USING gin (f_unaccent(title::text) gin_trgm_ops)
```

to be reflected properly.

Refs #7871 
cc @LACabeza

markstory committed:
Use postgres catalog to fetch column names.

Matching column names with regex is not reliable in postgres. Instead
use pg_catalog.attribute to access the correct column names. We need to
handle N rows for a single index as the array to string functions do not
preserve the column order in the index.

Refs #7871

markstory committed:
Extract a method for building constraints.

Refs #7871

lorenzo commented on src/Database/Schema/PostgresSchema.php:
```diff
@@ -229,46 +233,54 @@ public function convertIndexDescription(Table $table, $row)
         if ($row['indisunique'] && $type === Table::INDEX_INDEX) {
             $type = Table::CONSTRAINT_UNIQUE;
         }
-        preg_match('/\(([^\)]+?)\s*(?:ASC|DESC)?(?:NULLS FIRST|LAST)?\)/', $row['statement'], $matches);
```
I'm glad this is gone again :D

lorenzo commented:
Postgres tests failing

LACabeza commented:
With this sql, when the index is composed by two or more fields, it'll be fetched a row for each of the fields, like.

relname | attname | statement
-------  | ------- | ---------
facts_pkey | source | CREATE UNIQUE INDEX facts_pkey ON facts USING btree (source, code)
facts_pkey | code | CREATE UNIQUE INDEX facts_pkey ON facts USING btree (source, code)

Won't this be a problem? I saw on tests, this may be the problem for testSaveNewErrorCompositeKeyNoIncrement

markstory commented:
@LACabeza The subselect changes the order of the columns in the index which is also a problem as. That's why I went with the multiple row approach. There is clearly an issue to iron out around composite primary keys though.

markstory committed:
Clean up serial column detection.

Use more postgres catalog data to correctly identify serial columns.
This also means we can remove additional hacky code around guessing
serial columns.

Refs #7871

markstory commented:
I think I got this fixed now. I was able to purge out more hacky code too which is nice.

markstory commented on tests/TestCase/Database/Schema/PostgresSchemaTest.php:
```diff
@@ -288,7 +288,7 @@ public function testDescribeTable()
                 'unsigned' => null,
                 'comment' => null,
-                'autoIncrement' => true,
+                'autoIncrement' => false,
```
The schema definition for this table does not define this column as a serial, and instead only has `BIGINT PRIMARY KEY` which is not a serial type.

josegonzalez replied:
Does that affect how CakePHP will treat the column internally?

markstory replied:
Yes. But the current code results in a more accurate reflection of the schema.
